GoogleSignIn/Sources/Public/GoogleSignIn/GIDAppCheckProvider.h

```diff
@@ -0,0 +1,39 @@
+/*
+ * Copyright 2023 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#import <Foundation/Foundation.h>
+
+NS_ASSUME_NONNULL_BEGIN
+
+@class FIRAppCheckToken;
+
+#if TARGET_OS_IOS && !TARGET_OS_MACCATALYST
+
+NS_AVAILABLE_IOS(14)
+@protocol GIDAppCheckProvider <NSObject>
```

Comment:
Is there a reason why making this protocol is undesirable?

The trouble I have with using an initializer for composition is that `GIDSignIn` is currently a singleton, which makes using initializer composition tricky. Moreover, it would encourage and require writing tests that will depart from the actual API requirements of the class (i.e., unit tests would not go through the `GIDSignIn` shared singleton and would instead exercise a member-wise initializer that is only exposed internally.

Finally, this configuration method on `GIDSignIn` is using composition; admittedly, it's not initializer composition, but it is composition nonetheless.

```
- (void)configureWithAppCheckProvider:(nullable id<GIDAppCheckProvider>)provider
                           completion:(nullable void (^)(NSError * _Nullable error))completion
``` 

Last, I believe the `GIDAppCheck` does follow the pattern you describe: it has an initializer that takes an instance conforming to this protocol; if that instance is not nil, we will use that implementation for App Check work. Otherwise, we will use the conformance coming from Firebase (which we add ourselves elsewhere in this PR).

In summary, I think resolving this issue as you suggest would require a rewrite of `GIDSignIn`. Since that is beyond the scope of this PR, and is related to other ongoing work, I think we should hold off on these changes and pick them up when `GIDSignIn` is ready (i.e., when it is not a singleton).

Thoughts?

Reply:
As an SDK, exposing unnecessary types (or API surface in general) creates long-term maintenance burden. If we expose the `GIDAppCheckProvider` protocol, we may find app developers start to create their own fakes for their tests. And if we ever need to update the protocol, app developers would need to update their fakes, which is additional friction for them to use our latest SDK.

`GIDSignIn` already has a private initializer in GoogleSignIn/Sources/GIDSignIn_Private.h, you can add `id<GIDAppCheckProvider>` as an additional parameter to the initializer there.

We use this private initializer in our unit tests. The initialization of the shared singleton also calls through the private initializer so a decent level of consistency between tests and external use of the class is maintained.

Initializer composition and manual dependency injection through the initializer is a pattern we established in the `refactorGSI` branch. I think it's good to align our practices at this point so we don't need to refactor this code in refactorGSI again.